### PR TITLE
feat: conversation

### PR DESCRIPTION
**Pull request by ffflorian**

### CODE REVIEW

File: src/main/conversation/ConversationAPI.ts [hunk 1]
```diff
@@ -9,23 +22,264 @@ export default class ConversationAPI {
...
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Get all conversation IDs.
+   * @param limit Max. number of IDs to return
+   * @param conversationId Conversation ID to start from (exclusive)
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/conversationIds
+   */
+  public getConversationIds(limit: number, conversationId?: string): Promise<ConversationIds> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: Do we use this endpoint?

File: src/main/conversation/ConversationAPI.ts [hunk 2]
```diff
@@ -9,23 +22,264 @@ export default class ConversationAPI {
...
+
+    return _get_conversations();
+  }
+
+  /**
+   * Create a 1:1-conversation.
+   * @param conversationData The new conversation
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/createOne2OneConversation
+   */
+  public postOne2OneConversation(conversationData: NewConversation): Promise<{}> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: post1to1conversation is shorter

File: src/main/conversation/ConversationAPI.ts (hunk 1 above)
Comment (ffflorian) [reply src/main/conversation/ConversationAPI.ts]: No, but after coordination with @bennyn we decided to leave keep the APIs which are currently not in use.

File: src/main/conversation/Conversation.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+interface Conversation {
+  access?: any; // not used
+  creator: string;
+  id: string;
+  last_event: any; // not used
+  last_event_time: any; // not used
+  members: ConversationMembers;
+  name?: string;
+  team: any; // not used
```
Comment (bennyn) [inline src/main/conversation/Conversation.ts]: `team: string; // Legacy Team ID`

File: src/main/conversation/Conversation.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import {ConversationMembers} from '../conversation';
+
+interface Conversation {
+  access?: any; // not used
+  creator: string;
+  id: string;
+  last_event: any; // not used
```
Comment (bennyn) [inline src/main/conversation/Conversation.ts]: It's a `string`.

File: src/main/conversation/Conversation.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  last_event_time: any; // not used
```
Comment (bennyn) [inline src/main/conversation/Conversation.ts]: It's a `string` (ISO).

File: src/main/conversation/ConversationAPI.ts [hunk 6]
```diff
@@ -9,23 +22,264 @@ export default class ConversationAPI {
...
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Join a conversation.
+   * @param conversationId The conversation ID
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/joinConversation
+   */
+  public postJoinConversation(conversationId: string): Promise<any> {
```
Comment (bennyn) [inline src/main/conversation/ConversationAPI.ts]: `Promise<any>` is a no go.

File: src/main/conversation/Event.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+interface Event {
+  conversation: string;
+  time: string;
+  from: string;
+  type:
+    | 'conversation.member-join'
+    | 'conversation.member-leave'
```
Comment (bennyn) [inline src/main/conversation/Event.ts]: types can be sorted

File: src/main/conversation/Conversation.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
+import {ConversationMembers} from '../conversation';
+
+interface Conversation {
+  access?: any; // not used
```
Comment (bennyn) [inline src/main/conversation/Conversation.ts]: `access` is/was a `string`

File: src/main/conversation/Conversation.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+interface Conversation {
+  access?: string; // not used
+  creator: string;
+  id: string;
+  last_event_time: string; // not used
+  last_event: string; // not used
+  members: ConversationMembers;
+  name?: string;
+  team: string; // Legacy Team ID
```
Comment (gregor) [inline src/main/conversation/Conversation.ts]: This is not legacy and used in the new team API

File: src/main/conversation/ConversationAPI.ts [hunk 10]
```diff
@@ -9,23 +22,265 @@ export default class ConversationAPI {
...
+    };
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Recursively get all conversations.
+   * @param limit Max. number of conversations to return
+   */
+  public getAllConversations(limit: number = 500): Promise<Conversation[]> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: lets sort them alphabetically

File: src/main/conversation/ConversationAPI.ts [hunk 11]
```diff
@@ -9,23 +22,265 @@ export default class ConversationAPI {
...
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Join a conversation.
+   * @param conversationId The conversation ID
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/joinConversation
+   */
+  public postJoinConversation(conversationId: string): Promise<Event> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: `postConversation`

File: src/main/conversation/ConversationAPI.ts [hunk 12]
```diff
@@ -9,23 +22,265 @@ export default class ConversationAPI {
...
-        recipients: payload,
-      },
+      data: messageData,
       method: 'post',
-      url: `/conversations/${conversationId}/otr/messages?ignore_missing=${hasContent}`,
+      url: `${ConversationAPI.URL.CONVERSATIONS}/${conversationId}/${ConversationAPI.URL.OTR}/${ConversationAPI.URL
+        .MESSAGES}?ignore_missing=${hasContent}`,
     };
 
     return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: Seems like `addbot is missing`

File: src/main/conversation/ConversationAPI.ts (hunk 11 above)
Comment (ffflorian) [reply src/main/conversation/ConversationAPI.ts]: We already have `postConversation` for creating a new conversation. `postJoinConversation` joins a conversation.

File: src/main/conversation/ConversationAPI.ts (hunk 12 above)
Comment (ffflorian) [reply src/main/conversation/ConversationAPI.ts]: Yes, I could not find it in Swagger. Will add it manually.

Comment (gregor) [reply src/main/conversation/ConversationAPI.ts]: We use it in our webapp code. `postBots` and `deleteBots`

File: src/main/conversation/ConversationAPI.ts (hunk 11 above)
Comment (bennyn) [reply src/main/conversation/ConversationAPI.ts]: You can make it `postConversation` & `postConversationJoin`.

Comment (ffflorian) [reply src/main/conversation/ConversationAPI.ts]: :+1: good idea

File: src/main/conversation/ConversationAPI.ts [hunk 13]
```diff
@@ -1,33 +1,321 @@
...
+    };
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Recursively get all conversations.
+   * @param limit Max. number of conversations to return
+   */
+  public getAllConversations(limit: number = 500): Promise<Conversation[]> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: This is not for the API layer. A getAllConversations is not a backend endpoint and as suche not be handled in the API client but one layer above.

File: src/main/conversation/ConversationAPI.ts [hunk 14]
```diff
@@ -1,33 +1,321 @@
...
+
+    if (conversationId) {
+      config.url += `&start=${conversationId}`;
+    }
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Get all conversation
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: As the limits imply, it does not get all conversations  -> Get conversations

File: src/main/conversation/ConversationAPI.ts [hunk 15]
```diff
@@ -1,33 +1,321 @@
...
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Create a 1:1-conversation.
+   * @param conversationData The new conversation
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/createOne2OneConversation
+   */
+  public post1to1Conversation(conversationData: NewConversation): Promise<{}> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: from the naming it should be `post1to1`

File: src/main/conversation/ConversationAPI.ts [hunk 16]
```diff
@@ -1,33 +1,321 @@
...
+    };
+
+    return this.client.sendJSONRequest(config).then((response: AxiosResponse) => response.data);
+  }
+
+  /**
+   * Create a self-conversation.
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/createSelfConversation
+   */
+  public postSelfConversation(): Promise<Conversation> {
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: `postSelf`

File: src/main/conversation/ConversationAPI.ts [hunk 17]
```diff
@@ -1,33 +1,321 @@
...
+   * Send typing notifications.
+   * @param conversationId The Conversation ID
+   * @param typingData The typing status
+   * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/isTyping
+   */
+  public postTyping(conversationId: string, typingData: Typing): Promise<{}> {
+    const config = {
+      data: typingData,
+      method: 'post',
+      url: `${ConversationAPI.URL.CONVERSATIONS}/${conversationId}/self`,
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: URL is wrong. -> Add typing to constant and use here.

File: src/main/conversation/ConversationAPI.ts [hunk 18]
```diff
@@ -104,20 +81,23 @@ export default class ConversationAPI {
    * @see https://staging-nginz-https.zinfra.io/swagger-ui/#!/conversations/conversationIds
    */
   public getConversationIds(limit: number, conversationId?: string): Promise<ConversationIds> {
-    const config = {
+    const config: AxiosRequestConfig = {
+      data: {
+        size: limit,
+      },
       method: 'get',
       url: `${ConversationAPI.URL.CONVERSATIONS}/ids?size=${limit}`,
```
Comment (gregor) [inline src/main/conversation/ConversationAPI.ts]: Remove from here now that you added it to data

Comment (ffflorian) [reply src/main/conversation/ConversationAPI.ts]: I have to revert it to query string because of https://github.com/mzabriskie/axios/issues/462.

**Pull request merged**